File: Dotnet/dotnet-core-sample/dotnet-console-sample/Postgre/PostgreBase.cs
using Microsoft.Extensions.Configuration;
using Npgsql;
using NpgsqlTypes;
using Serilog;
using UUIDNext;

namespace dotnet_console_sample.Postgre
{
	// Postgre的基础示例
	class PostgreBase
	{
		// appsettings.json 和 环境变量
		private readonly IConfiguration _config;

		public PostgreBase(IConfiguration config)
		{
			_config = config;
		}
		// 程序入口
		public async Task PostgreBaseStart()
		{
			// 需要1: dotnet add package Npgsql
			// 需要2: docker run -p 5432:5432 --name postgre -v $HOME/workspace/postgre_data/data:/var/lib/postgresql/data -e POSTGRES_PASSWORD=123456 -d postgres:17.1-alpine3.20
			// 需要3: 建表
			/*
			CREATE TABLE t_employee (
				te_pk UUID NOT NULL UNIQUE,
				employee_id TEXT NOT NULL,
				employe_name TEXT,
				employe_email TEXT,
				employe_status SMALLINT,
				PRIMARY KEY(te_pk)
			);
			*/
			// 更多看这里: https://www.npgsql.org/doc/index.html

			string? host = _config["PostgreSettings:Host"];
			string? port = _config["PostgreSettings:Port"];
			string? database = _config["PostgreSettings:Database"];
			string? username = _config["PostgreSettings:Username"];
			string? password = _config["PostgreSettings:Password"];
			string connString = $"Host={host};Port={port};Username={username};Password={password};Database={database}";
			Log.Information("Postgre连接URL: {connString}", connString);

			// 连接数据库
			// 默认情况下，Npgsql 连接是池化的：关闭或处置连接不会关闭底层物理连接，而是将其返回到由 Npgsql 管理的内部池
			var dataSourceBuilder = new NpgsqlDataSourceBuilder(connString);
			var dataSource = dataSourceBuilder.Build();
			await using var connection = await dataSource.OpenConnectionAsync();
			Log.Information("Postgre连接成功");

			// 插入3条t_employee数据
			for (int i = 0; i < 3; i++)
			{
				// 插入t_employee数据
				await using (var cmd = new NpgsqlCommand("INSERT INTO t_employee (te_pk, employee_id, employe_name, employe_status) VALUES (@te_pk, @employee_id, @employe_name, @employe_status)", connection))
				{
					// 创建UUID v7版本的主键
					string uuidv7 = Uuid.NewDatabaseFriendly(Database.PostgreSql).ToString();
					Guid te_pk = Guid.Parse(uuidv7);
					// 取得1到99的随机数
					int randomNumber = new Random().Next(1, 100);
					string employee_id = "dotnet_base_id_" + randomNumber;
					string employe_name = "dotnet_张三_" + randomNumber;
					int employe_status = 1;
					cmd.Parameters.AddWithValue("te_pk", NpgsqlDbType.Uuid, te_pk);
					cmd.Parameters.AddWithValue("employee_id", employee_id);
					cmd.Parameters.AddWithValue("employe_name", employe_name);
					cmd.Parameters.AddWithValue("employe_status", employe_status);
					await cmd.ExecuteNonQueryAsync();
					Log.Information("t_employee表插入成功, {te_pk}, {employee_id}, {employe_name}, {employe_status}", uuidv7, employee_id, employe_name, employe_status);
				}
			}

			// 查询t_employee所有数据
			// 创建一个List，其中每个元素都是Dictionary<string, string>类型
			List<Dictionary<string, string>> results = [];
			await using (var cmd = new NpgsqlCommand("SELECT * FROM t_employee", connection))
			await using (var reader = await cmd.ExecuteReaderAsync())
			{
				while (await reader.ReadAsync())
				{
					string te_pk = reader.GetFieldValue<Guid>(0).ToString();
					string employee_id = reader.GetString(1);
					string employe_name = reader.GetString(2);
					string? employe_email = reader.IsDBNull(reader.GetOrdinal("employe_email")) ? null : reader.GetString(reader.GetOrdinal("employe_email"));
					if (string.IsNullOrEmpty(employe_email))
					{
						employe_email = "N/A";
					}
					int employe_status = reader.GetInt16(4);
					// 创建字典并添加到List中
					Dictionary<string, string> record = new()
					{
						{ "te_pk", te_pk },
						{ "employee_id", employee_id },
						{ "employe_name", employe_name },
						{ "employe_email", employe_email },
						{ "employe_status", employe_status.ToString() },
					};
					results.Add(record);
				}
			}
			Log.Information("t_employee表全部数据查询成功");
			Log.Information("{results}", results);

			// 关闭数据库连接
			connection.Close();
		}
	}
}

File: Go/GoSampleProject/src/postgre/postgre_base.go
package postgre

import (
	"container/list"
	"context"
	"fmt"
	"github.com/google/uuid"
	"github.com/jackc/pgx/v5/pgxpool"
	"math/rand"
	"my.sample/GoSampleProject/src/config"
	"my.sample/GoSampleProject/src/log"
)

// Postgre的基础示例
func PostgreBase() {
	// 需要1:
	// go get github.com/jackc/pgx/v5  (单连接模式，不是并发安全)
	// go get github.com/jackc/pgx/v5/pgxpool  (连接池模式，并发安全)
	// 需要2: docker run -p 5432:5432 --name postgre -v $HOME/workspace/postgre_data/data:/var/lib/postgresql/data -e POSTGRES_PASSWORD=123456 -d postgres:17.1-alpine3.20
	// 需要3: 建表
	/*
		CREATE TABLE t_employee (
			te_pk UUID NOT NULL UNIQUE,
			employee_id TEXT NOT NULL,
			employe_name TEXT,
			employe_email TEXT,
			employe_status SMALLINT,
			PRIMARY KEY(te_pk)
		);
	*/
	// 更多看这里:
	// https://github.com/jackc/pgx
	// https://github.com/jackc/pgx/blob/master/examples/todo/main.go

	ctx := context.Background()

	// 从config.yaml文件读取数据库连接信息
	// dbUrl := os.Getenv("DATABASE_URL")
	dbUrl := config.Viper.GetString("datasource.url")
	log.Logger.Info().Msgf("Postgre连接URL: %v", dbUrl)
	// 创建数据库连接池
	dbpool, err := pgxpool.New(ctx, dbUrl)
	if err != nil {
		log.Logger.Error().Stack().Err(err).Msgf("Postgre连接失败: %v", err)
		panic(err)
		// os.Exit(1)
	}
	defer func() {
		dbpool.Close()
		log.Logger.Info().Msg("Postgre连接池已关闭")
	}()

	// 测试连接
	err = dbpool.Ping(ctx)
	if err != nil {
		log.Logger.Error().Stack().Err(err).Msgf("Postgre连接失败: %v", err)
		panic(err)
	} else {
		log.Logger.Info().Msg("Postgre连接成功")
	}

	// 插入3条t_employee数据
	for i := 0; i < 3; i++ {
		// 对数据库友好的 UUID v7
		uuidv7, _ := uuid.NewV7()
		tepk := uuidv7.String()
		// 生成1到99的随机数
		randomNumber := rand.Intn(99) + 1
		employee_id := fmt.Sprintf("%s%d", "golang_base_id_", randomNumber)
		employe_name := fmt.Sprintf("%s%d", "dotnet_张二_", randomNumber)
		employe_status := 3
		_, err = dbpool.Exec(ctx, "INSERT INTO t_employee (te_pk, employee_id, employe_name, employe_status) VALUES ($1, $2, $3, $4)", tepk, employee_id, employe_name, employe_status)
		if err != nil {
			log.Logger.Error().Stack().Err(err).Msgf("插入t_employee失败: %v", err)
			panic(err)
		}
		log.Logger.Info().Msgf("t_employee表插入成功, %v, %v, %v, %v", tepk, employee_id, employe_name, employe_status)
	}

	// 查询t_employee所有数据
	// 创建一个List，其中每个元素都是map<string, string>类型
	results := list.New()
	rows, err := dbpool.Query(ctx, "SELECT * FROM t_employee")
	for rows.Next() {
		var te_pk string
		var employee_id string
		var employe_name string
		var employe_email *string
		var employe_status int
		err := rows.Scan(&te_pk, &employee_id, &employe_name, &employe_email, &employe_status)
		if err != nil {
			log.Logger.Error().Stack().Err(err).Msgf("查询t_employee失败: %v", err)
			panic(err)
		}
		if employe_email == nil {
			temp := "N/A"
			employe_email = &temp
		}
		// 创建map并添加到List中
		record := make(map[string]string)
		record["te_pk"] = te_pk
		record["employee_id"] = employee_id
		record["employe_name"] = employe_name
		record["employe_email"] = *employe_email
		record["employe_status"] = fmt.Sprintf("%d", employe_status)
		results.PushBack(record)
	}
	if err != nil {
		log.Logger.Error().Stack().Err(err).Msgf("查询t_employee失败: %v", err)
		panic(err)
	}
	log.Logger.Info().Msgf("t_employee表查询成功")
	// 遍历list并打印其中的map元素
	for element := results.Front(); element != nil; element = element.Next() {
		record := element.Value.(map[string]string)
		log.Logger.Info().Msgf("%v", record)
	}
}
